### ISSUE TITLE
Add realms logout support to internal_api for native

### ISSUE THREAD

**Issue by jedp** [mozilla/persona#3915]
#3854 provides realms support to Persona.  For the ability to logout of a realm on FirefoxOS, I think we need to do a little platform work.

[Bug 919157](https://bugzilla.mozilla.org/show_bug.cgi?id=919157) Tracks the necessary client code for FirefoxOS.

The soonest realms logout support can arrive in FirefoxOS is version 1.3, which is feature-complete in December 2013 and goes to market in March 2014.  I can't see any way to make prior versions of FirefoxOS compatible with realms purely from the server side, so I think we're just stuck with this.

**Comment by seanmonstar** [comment]
I'd love to see if we can get it working with current code. I don't see anything special that I'm doing to logout. Just looping an array and removing `logged_in` on sites. Is this special cased in Fx?

**Comment by jedp** [comment]
It's special-cased in FirefoxOS because people didn't want the Persona iframe to be persistent and take up the 10MB that an app eats up just by existing (each app is a full-on gecko process). So the persona iframe comes and goes as needed for DOM API calls.  Here is the [epic bug](https://bugzilla.mozilla.org/show_bug.cgi?id=839500) behind the current implementation.

Obviously, this creates a problem for us because the dialog can no longer communicate back to the RPs on the other side of the fence.  So for logout to work correctly, we had to implement in the client the iteration through RPs with the same origin: https://mxr.mozilla.org/mozilla-central/source/toolkit/identity/MinimalIdentity.jsm#245

I don't see an alternative but to do the same thing for realms :(

**Comment by seanmonstar** [comment]
That's called when a site calls navigator.id.logout()? So
BrowserID.internal.logout() isn't used?

**Comment by lloyd** [comment]
one question, we're actively exploring a new api which moves away from full session support - specifically it would never fire onlogout events.  

How much would this simplify our world here?

**Comment by callahad** [comment]
I believe realms must have full-fledged session support, login and logout included.

A and B are members of Realm.
1. I manually log into A as Foo.
2. I navigate to B and am automatically logged in as Foo. (So far so good!)
3. I navigate to A and manually log out.
4. I navigate to B.

Without automatic onlogout events, I'm still logged in at B, which is Not Correct.

**Comment by callahad** [comment]
(Tangentially related: As per #3916, Safari on Desktop is also inhibiting localStorage by default as of 6.1, which limits our ability to do full-fledged session stuff there. Boo.)

**Comment by lloyd** [comment]
@callahad I'm confused.  How can we consider a feature that "must have full session support" and an API that eliminates logout at the same time?

**Comment by seanmonstar** [comment]
Well, we have to keep the current session based api. We promised it's
stability when we announced Beta 1. It's in use on sites.

I assumed were going to make an adjust api without session management, and
recommend that in most circumstances. But if you want realms, you'll need
session management.

Or, we can offer realms with a lesser experience.

**Comment by callahad** [comment]
@lloyd My point is that anything less than Full-On-Observer-API is insufficient for realms / cross-domain SSO. The Paris Proposal doesn't buy us much there.

I'm with Sean. Realms are a super advanced topic. Let's have a "Basic," session-free API. If you want sophisticated features, you have to upgrade to a more sophisticated API. I'm fine with that, so long as that complexity is opt-in.

**Comment by lloyd** [comment]
Our core premise in the past has been let's not bifurcate the api into a simple and a complex version - we tried this once and found it didn't work.

We're conflating two things here necessarily because they affect each other.

The first is the question - can we make a decisive change to our api that will be address the majority of the problems with the .watch() api - we aim to sacrifice a feature to improve performance and simplicity.

If we answer the first question, this patch becomes 100 lines (emailHint) instead of 900 (requiredEmail).

Don't get me wrong, I love the feature, I'm just saying drawing the "advanced usage" line in a different place.  I'm saying muti-site login is basic, and simultaneous logout is advanced.  Sites who want this can implement it themselves.  Or they can simply use session cookies and session prologation.  and .logout() on one site in a realm can modify localstorage and remove the logged in state for all.

Bottom Line - I'm uncomfortable landing this patch before we make a decision on our API.

**Comment by callahad** [comment]
[Whoops, mis-click. Sorry about that.]

I am under the impression that SSO which does not fully synchronize state is not acceptable by the RPs requesting this feature. I can check back in with them and validate that assumption.

> a simple and a complex version - we tried this once and found it didn't work.

Another way to view this: We found that our simple API worked. It matched developers' understanding, it was simple, and it got implemented correctly. The only intrinsic flaw with .get was that the callback wasn't visible to the browser, so we couldn't close the loop after an email verification or redirect.

We know that works.

We don't yet know if/how a session-managed API works, and we're still actively exploring that space. Let's standardize a simple, .get-like API _now,_ and try to extend it into session management in the future, rather than committing to both right away. That lets us unblock PICL / Native, and buys us time to experiment with session management.

...I'll move this to the mailing list.

**Comment by jaredhirsch** [comment]
I'm concerned about the amount of roadmappy shit happening in this bug. Can somebody summarize (when this finally finishes) and take it to the list, so we have a single source of truth for the realms conversations? :-\
